src/use-transition/use-transition.ts

```diff
@@ -0,0 +1,154 @@
+
+interface TransitionController extends Controller {
+  show: (event: Event) => void
+  hide: (event: Event) => void
+  toggle: (event: Event) => void
+  isOpen: boolean
+}
+
+export interface TransitionOptions {
+  element?: Element
+  dispatchEvent?: boolean
+  open?: boolean
+  eventPrefix?: boolean | string
+  enter?: string
+  enterActive?: string
+  enterTo?: string
+  leave?: string
+  leaveActive?: string
+  leaveTo?: string
+  hiddenClass?: string
+}
+
+const alpineNames: object = {
+  enterClass: "enter",
+  enterActiveClass: "enterStart",
+  enterToClass: "enterEnd",
+  leaveClass: "leave",
+  leaveActiveClass: "leaveStart",
+  leaveToClass: "leaveEnd",
+}
+
+const defaultOptions = {
+  dispatchEvent: true,
+  eventPrefix: true,
+  open: false,
+  hiddenClass: "hidden"
+}
+
+export const useTransition = (controller: TransitionController, options: TransitionOptions = {}) => {
+  const { open, hiddenClass } = Object.assign(defaultOptions, options)
+  const targetElement = options?.element || controller.element
+  if (!((targetElement instanceof HTMLElement) || (targetElement instanceof SVGElement))) return
+  const dataset = targetElement.dataset
+
+  const controllerShow: Function = controller.show?.bind(controller)
+  const controllerHide: Function = controller.hide?.bind(controller)
+  const controllerToggle: Function = controller.toggle?.bind(controller)
+
+  async function show(event?: Event) {
+    if (controller.isOpen) return
+
+    controller.isOpen = true
+    controllerShow && controllerShow(event)
+
+    const enterClass = getAttribute("enter", options, dataset)
+    const enterActiveClass = getAttribute("enterActive", options, dataset)
+    const enterToClass = getAttribute("enterTo", options, dataset)
+
+    if (!!hiddenClass) {
+      targetElement.classList.remove(hiddenClass)
+    }
+
+    await transition(targetElement, enterClass, enterActiveClass, enterToClass)
+
+  }
+
+  async function hide(event?: Event) {
+    if (!controller.isOpen) return
+
+    controller.isOpen = false
+    controllerHide && controllerHide(event)
+
+    const leaveClass = getAttribute("leave", options, dataset)
+    const leaveActiveClass = getAttribute("leaveActive", options, dataset)
+    const leaveToClass = getAttribute("leaveTo", options, dataset)
+
+    await transition(targetElement, leaveClass, leaveActiveClass, leaveToClass)
+
+    if (!!hiddenClass) {
+      targetElement.classList.add(hiddenClass)
+    }
+  }
+
+  function toggle(event: Event) {
+    controllerToggle && controllerToggle(event)
+    if (controller.isOpen) {
+      hide()
+    } else {
+      show()
+    }
+  }
+
+  async function transition(element: Element, initialClass: string[], activeClass: string[], toClass: string[]) {
+    element.classList.add(...initialClass);
+    element.classList.add(...activeClass);
+    await nextFrame();
+
+    element.classList.remove(...activeClass);
```

Comment:
Something I ran into when implementing transitions was there were a few cases where I had a class on the element initially, but it was also in the active transition. I started with a line exactly like this, but that meant when the transition finished the class that was initially on the element was removed as well. Alpine handles this by storing the initial classes and leaving those in place when removing the active classes https://github.com/alpinejs/alpine/blob/master/src/utils.js#L427

If you're interested the package I have been using is [here](https://github.com/mmccall10/el-transition) and the patch for this issue [here](https://github.com/mmccall10/el-transition/pull/2). Your approach looks very similar, looking forward to switching over to your version and dropping a dependency!

Reply:
Thanks @jduff I started to feel similar issues yesterday while testing it. 
I agree that this should be addressed. Will look a bit more into Alpine code to make sure we have something equivalent in terms of edges cases.